src/main/scala/linting/ChiselLinting.scala

```diff
@@ -0,0 +1,122 @@
+package freechips.rocketchip.linting
+
+import firrtl._
+import firrtl.ir._
+import firrtl.traversals.Foreachers._
+import firrtl.annotations.NoTargetAnnotation
+import firrtl.options.{OptionsException, RegisteredLibrary, ShellOption}
+import firrtl.stage.RunFirrtlTransformAnnotation
+import chisel3.experimental.ChiselAnnotation
+
+
+import scala.collection.mutable
+
+case class LintError(info: FileInfo, modules: Set[String])
+case class LintExceptions(seq: Seq[LintError]) extends FirrtlUserException(
+  s"""Linting Report: ${seq.size} exceptions!
+     |Please fix using @chiselName, *.suggestName(...). Otherwise, whitelist the file via the Chisel commandline option:
+     |    --linting-whitelist <filename>.scala
+     |or via annotations in your Chisel code:
+     |    annotate(LintWhitelist("<filename>.scala")).
+     |
+     |Only showing first few exceptions:
+     |${seq.zip(0 until 20)
+           .map { case (lint: LintError, idx: Int) => s"$idx. ${lint.info} in ${lint.modules}"}
+           .mkString("\n")}
+     """.stripMargin
+)
+
+case class LintWhitelist(whiteList: Set[String]) extends NoTargetAnnotation with ChiselAnnotation {
+  override def toFirrtl = this
+}
+
+final class ChiselLinting extends Transform with RegisteredLibrary {
```

Comment:
Moving my feedback from generic comment to here so it can be discussed in one spot:

I would be more specific than just "linting" since there are other possible lint rules we may want to add some day. I would call this `RegNameLinting` or something like that.

Reply:
So I imagine we could have one ChiselLinting transform, which does multiple linting passes internally, no?